### PR TITLE
Route Viewer Search and Filter

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
This PR adds search and filter capabilities to the route viewer. It also moves the route viewer to the Navbar to make it easier to find.

### CODE REVIEW

File: lib/components/app/desktop-nav.js
Comment (landonreed) [inline lib/components/app/desktop-nav.js:57]: I like where your heads at here, but I think we need to preserve this space for agency branding. Is there a possibility to stick this switcher at the bottom of the page (both mobile and desktop)? Before implementing this change, let's get consensus from the team.

File: lib/components/app/view-switcher.js
Comment (landonreed) [inline lib/components/app/view-switcher.js:64]: Might be worth swapping this for react-intl approach.

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:105]: Can this `modes` variable filter the available options based on only the routes for the presently selected agency? E.g., if Cobb is selected, only show bus routes (rather than bus, subway, and tram).

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:156]: Let's go ahead and add react-intl (everywhere in this file) if you can coordinate/get the OK from @binh-dam-ibigroup.

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:42]: Slight pref for this stuff to live in redux, and leverage a selector to get the list of routes (rather than doing the filtering within the render method).

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:135]: For this section, I think the mobile view should be how things look on desktop too. I.e., two rows (top with selects, bottom with text search). Also, can we extend the agency filter to fill the empty space (since oftentimes, the agency names are quite long and get cut off)?
image

**Review by landonreed — CHANGES_REQUESTED** [review]
Love this! Just a few changes.

File: lib/components/app/desktop-nav.js
Comment (miles-grant-ibigroup) [reply lib/components/app/desktop-nav.js:57]: Are we expecting branding to get wider than this? Even wide-ish branding fits nicely. But agree good to discuss with everyone.

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Handled in e886cfc75dde3dbec1844daa3cd88c51a0750e23

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:43]: Moved to redux in e886cfc75dde3dbec1844daa3cd88c51a0750e23

File: lib/components/viewers/route-viewer.js
Comment (binh-dam-ibigroup) [reply lib/components/viewers/route-viewer.js:138]: @miles-grant-ibigroup For the OTP mode of each route (e.g. `BUS`, `SUBWAY`, etc), you can access the localized strings using the message id `common.otpTransitModes.${mode.toLowerCase}`.

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:116]: All possible and all done in f638c7f23bedaa87c1472472090145f129563cca

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Internationalized in ae4bf350860169e3e74de8c5596047fdb3add159

File: lib/components/app/view-switcher.js
Comment (miles-grant-ibigroup) [reply lib/components/app/view-switcher.js]: Swapped in b046f097e7325e2a6e5d363bef7c696c4d69fca4

**Comment by miles-grant-ibigroup** [comment]
Everything addressed! With the redux I did my best to follow conventions, but am very open to restructuring things to match the rest of the project. Also I'd like to propose @binh-dam-ibigroup as we're adding `React-intl` stuff

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:75]: I feel like there really shouldn't be much lag there... maybe there are some perf improvements we should target.

Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Fixed this issue in e886cfc75dde3dbec1844daa3cd88c51a0750e23 (line 56 of route-viewer.js) and removed react state in 428ff8bd42e0757e6967d56f15b19efe0d8f63b3. Still runs smooth!

File: lib/components/viewers/route-viewer.js
Comment (landonreed) [inline lib/components/viewers/route-viewer.js:90]: Can this sort functionality live in the selector, too?

Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: It very much can! Moved in ec7e88565eba513e9799ecc01296f122ac9a937f

**Comment by miles-grant-ibigroup** [comment]
Thanks Landon! Changes are addressed.

File: lib/components/app/desktop-nav.js
Comment (miles-grant-ibigroup) [reply lib/components/app/desktop-nav.js:57]: As per discussion, it is now shown on desktop but mobile retains current behavior. Changed in 2b85bbc484d641323dcc885541d14ac5b7e03285

File: lib/components/app/app-menu.js
Comment (miles-grant-ibigroup) [reply lib/components/app/app-menu.js:66]: Comment added and titles re-added in 85bcb56a190f9eef583ab29039f7941c6b7f0b9b

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Good catch thank you! It absolutely should be. Added in 511c2427322217d4d8e367ced31c241c34a7a6ea

**Comment by miles-grant-ibigroup** [comment]
Thanks! Re-added the back button and added other requested changes.

**Review by landonreed — APPROVED** [review]
This is so great, @miles-grant-ibigroup!

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Looks mostly good! Please address the following:
- Overriding the branding logo on mobile views,
- Displaying the `ViewSwitcher` on desktop screens less than 800px wide.
- One string needs to be localized.
- Various formatting comments.

Other items that are not blocking:
- Potentially using bootstrap `Nav`/`NavItem` elements in the `ViewSwitcher`
- Consistent behavior of the Back button across Route Viewer, Stop Viewer, etc.
- Sorting agencies according to `transitOperators` in config if available

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:285]: (nitpick) This is probably causing the mode filter selector to not be flush with the right margin.

File: lib/components/viewers/stop-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-viewer.js:55]: Yes. Handled in https://github.com/opentripplanner/otp-react-redux/pull/441/files#diff-e2fa18254c2006832bc83b8ae61e382698f19a4de1e7fb4895a71151fa74ff42R56

File: lib/components/app/desktop-nav.js
Comment (miles-grant-ibigroup) [reply lib/components/app/desktop-nav.js:46]: When this is removed the hamburger button is below the view switcher and can't be clicked.

File: i18n/en-US.yml
Comment (miles-grant-ibigroup) [reply i18n/en-US.yml]: If the space is removed react-intl just renders the string ID. Have added a comment in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/app/view-switcher.js
Comment (miles-grant-ibigroup) [reply lib/components/app/view-switcher.js]: Good catch thank you! Removed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/app/desktop-nav.js
Comment (miles-grant-ibigroup) [reply lib/components/app/desktop-nav.js:58]: Ah thank you! Have changed the method of hiding the switcher in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/app/view-switcher.js
Comment (miles-grant-ibigroup) [reply lib/components/app/view-switcher.js]: Have tried this, but it seemed silly adding a large number of css rules to remove the existing bootstrap styles. However, have changed from a `div` to a `nav` in a981abe4dac16acfc816cf54aa4b975ff102118f to make things a bit more semantically accurate.

File: lib/components/mobile/navigation-bar.js
Comment (miles-grant-ibigroup) [reply lib/components/mobile/navigation-bar.js]: Good catch thank you! Just forgot to remove this when refactoring ViewSwitcher to only appear on desktop. Fixed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:42]: Weird! Fixed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:136]: Yes it does! And it's required to make the filtering work properly.

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:171]: Good catch thank you! Removed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Should've been `aria-label`. Fixed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:328]: Fixed in https://github.com/opentripplanner/otp-react-redux/pull/441/files#diff-181a258a88e27648c35bb97f1dc995993979e4214458e1775698b8ae290b0725R157

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js:317]: This might be a question for the technical call. I would argue that even if using a different language the number and headsign should match the format used on the vehicle?

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: I agree, but can't see a method in core-utils that's exposed that can assist in doing this. Is there one I missed?

File: lib/components/app/desktop-nav.js
Comment (miles-grant-ibigroup) [reply lib/components/app/desktop-nav.js:47]: Comment added in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/util/state.js
Comment (miles-grant-ibigroup) [reply lib/util/state.js]: Removed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/util/state.js
Comment (miles-grant-ibigroup) [reply lib/util/state.js]: Removed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/viewers.css
Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css]: Thank you! That had been bothering me. Fixed in a981abe4dac16acfc816cf54aa4b975ff102118f

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:314]: Set to 10px.

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:287]: Is this one useful?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Looks good with some minor tweaks, but there are some changes I requested before that didn't make it, so please add them.

File: lib/components/app/app.css
Comment (miles-grant-ibigroup) [reply lib/components/app/app.css:74]: Fixed in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa

File: lib/components/app/view-switcher.js
Comment (miles-grant-ibigroup) [reply lib/components/app/view-switcher.js:14]: Added in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa

File: lib/components/app/view-switcher.js
Comment (miles-grant-ibigroup) [reply lib/components/app/view-switcher.js]: Fixed in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Was the result of a merge. Fixed in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa

File: lib/components/viewers/viewers.css
Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css]: Good catch thanks! Fixed in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa

File: lib/components/viewers/viewers.css
Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css:287]: Yes otherwise the dynamic spacing cuts the text

**Comment by miles-grant-ibigroup** [comment]
I can't respond to some of the comments, but most have been addressed in 3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa aside from:

- the localization comment, which is addressed in #441 which is set to be merged into this branch.
- the `<option value={''}>` which is needed based on how the reducer is set up

**Comment by binh-dam-ibigroup** [comment]
> Thanks for the translations and notes! I can't respond to some of the comments, but most have been addressed in [3bcaf13](https://github.com/opentripplanner/otp-react-redux/commit/3bcaf1302c8766d59aa3eb92f837832bb8ce3eaa) aside from:
> 
> * the localization comment, which is addressed in [Route Viewer Details Pane and Realtime Vehicle Position Viewer #441](https://github.com/opentripplanner/otp-react-redux/pull/441) which is set to be merged into this branch.
> * the `<option value={''}>` which is needed based on how the reducer is set up

Okay, thanks for clarifying.

File: lib/components/viewers/route-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/route-viewer.js:20]: Move this up (a-z)

File: lib/components/viewers/route-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/route-viewer.js:258]: Remove semicolon

**Review by binh-dam-ibigroup — APPROVED** [review]
Two more formatting things, but otherwise great work, thanks for the changes!

**Comment by miles-grant-ibigroup** [comment]
Thanks Binh! This is now blocked waiting for piggyback PR #441 to be merged

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Done in 59bdf8f495ea26a3a31f7b48ad10a767b0ca8ef7

File: lib/components/viewers/route-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/route-viewer.js]: Done in 59bdf8f495ea26a3a31f7b48ad10a767b0ca8ef7

File: lib/components/app/app.css
Comment (binh-dam-ibigroup) [reply lib/components/app/app.css:74]: Sorry I'm reversing my opinion on that... The outline is defined by bootstrap and is needed for keyboard navigation (tab, etc). There might be a way to remove it but support is not consistent, see https://css-tricks.com/keyboard-only-focus-styles/

Comment (miles-grant-ibigroup) [reply lib/components/app/app.css:74]: Have restored the outline, but adjusted the border radius to make it look more natural in a922e8f4e08da7d0f427298c6e43238b5a0b51e4

**Pull request merged**
